control-plane/cni/main.go

```diff
@@ -0,0 +1,223 @@
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// This is a sample chained plugin that supports multiple CNI versions. It
+// parses prevResult according to the cniVersion
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net"
+	"path/filepath"
+
+	"github.com/containernetworking/cni/pkg/skel"
+	"github.com/containernetworking/cni/pkg/types"
+	current "github.com/containernetworking/cni/pkg/types/100"
+	"github.com/containernetworking/cni/pkg/version"
+	"github.com/hashicorp/go-hclog"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/clientcmd"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	bv "github.com/containernetworking/plugins/pkg/utils/buildversion"
+)
+
+const (
+	keyCNIStatus    = "consul.hashicorp.com/cni-status"
+	keyInjectStatus = "consul.hashicorp.com/connect-inject-status"
+	injected        = "injected"
+)
+
+type CNIArgs struct {
+	types.CommonArgs
+	IP                         net.IP
+	K8S_POD_NAME               types.UnmarshallableString
+	K8S_POD_NAMESPACE          types.UnmarshallableString
+	K8S_POD_INFRA_CONTAINER_ID types.UnmarshallableString
+}
+
+// PluginConf is whatever you expect your configuration json to be. This is whatever
+// is passed in on stdin. Your plugin may wish to expose its functionality via
+// runtime args, see CONVENTIONS.md in the CNI spec.
+type PluginConf struct {
+	// This embeds the standard NetConf structure which allows your plugin
+	// to more easily parse standard fields like Name, Type, CNIVersion,
+	// and PrevResult.
+	types.NetConf
+
+	RuntimeConfig *struct {
+		SampleConfig map[string]interface{} `json:"sample_config"`
+	} `json:"runtime_config"`
+
+	Name string `json:"name"`
+	// Type of plugin (consul-cni)
+	Type string `json:"type"`
+	// CNIBinDir is the location of the cni config files on the node. Can bet as a cli flag.
+	CNIBinDir string `json:"cni_bin_dir"`
+	// CNINetDir is the locaion of the cni plugin on the node. Can be set as a cli flag.
+	CNINetDir string `json:"cni_net_dir"`
+	// Multus is if the plugin is a multus plugin. Can be set as a cli flag.
+	Multus bool `json:"multus"`
+	// Kubeconfig file name. Can be set as a cli flag.
+	Kubeconfig string `json:"kubeconfig"`
+	// LogLevl is the logging level. Can be set as a cli flag.
+	LogLevel string `json:"log_level"`
+}
+
+// parseConfig parses the supplied configuration (and prevResult) from stdin.
+func parseConfig(stdin []byte) (*PluginConf, error) {
```

Comment:
This is standard libncni code and is part of their sample. Libcni does a lot of the work for you when it comes to dealing with all of the structure around CNI plugins. Parsing the inputs being one of them...